zebra-script/src/lib.rs

```diff
@@ -87,9 +100,26 @@ impl CachedFfiTransaction {
             .expect("serialized transaction lengths are much less than u32::MAX");
         let mut err = 0;
 
-        // SAFETY: the `tx_to_*` fields are created from a valid Rust `Vec`.
+        let serialized_all_previous_outputs = all_previous_outputs
+            .zcash_serialize_to_vec()
+            .expect("serialization into a vec is infallible");
+        // TODO: pass to zcash_script after API update
+        let _all_previous_outputs_ptr = serialized_all_previous_outputs.as_ptr();
+        let _all_previous_outputs_len: u32 = serialized_all_previous_outputs
+            .len()
+            .try_into()
+            .expect("serialized transaction lengths are much less than u32::MAX");
+
+        // SAFETY:
+        // the `tx_to_*` fields are created from a valid Rust `Vec`
+        // the `all_previous_outputs_*` fields are created from a valid Rust `Vec`
         let precomputed = unsafe {
-            zcash_script::zcash_script_new_precomputed_tx(tx_to_ptr, tx_to_len, &mut err)
+            zcash_script::zcash_script_new_precomputed_tx(
+                tx_to_ptr, tx_to_len,
+                // all_previous_outputs_ptr,
```

Comment:
Note: I'm a bit worried if `CachedFfiTransaction` will still be safe to be `Send` when this is uncommented. We should be sure that the pointer is not stored inside the returned `precomputed` variable.

Maybe we should add a comment here?

```suggestion
                // TODO: Ensure that `all_previous_outputs_ptr` is not stored inside `precomputed`
                // so that `CachedFfiTransaction` is still safe to implement `Send`.
                // all_previous_outputs_ptr,
```

Reply:
This is a kinda outside our control, since `zcash_script` is not maintained by us. And we have the same risk with the `tx_to_ptr` buffer. It should be fine, since the precomputation is basically hashing a bunch of parts and caching the result, so it woun't need a reference to the previous outputs anymore (or the current tx). I'm not sure if we should handle this risk somehow (the only thing I can come up with is having a checklist for  `zcash_script` updates, but seems tricky to remember to use it). Or we could each out to ECC to suggest adding comment in the `zcash_script_new_precomputed_tx` header promising that no reference is kept to the passed buffers.

C is hard :smiling_face_with_tear:

Reply:
> Or we could each out to ECC to suggest adding comment in the `zcash_script_new_precomputed_tx` header promising that no reference is kept to the passed buffers.

Yes please file a PR against zcash_script.

Reply:
> > Or we could each out to ECC to suggest adding comment in the `zcash_script_new_precomputed_tx` header promising that no reference is kept to the passed buffers.
> 
> Yes please file a PR for this against `zcash_script`.

Let's add this to the list of tasks in our `zcash_script` API ticket #3291?

Reply:
> Note: I'm a bit worried if `CachedFfiTransaction` will still be safe to be `Send` when this is uncommented. We should be sure that the pointer is not stored inside the returned `precomputed` variable.

If `precomputed` stores and dereferences `tx_to_ptr` or `all_previous_outputs_ptr`, it will unconditionally invoke undefined behavior. Because those pointers are dropped at the end of `CachedFfiTransaction::new`.

So there is no interaction with `Send` or `Sync` here, because any use of those buffers is invalid, regardless of synchronization.

Reply:
@daira created PR https://github.com/zcash/zcash/pull/5504